File: 0113-path-sum-ii/0113-path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public void pathSum(TreeNode root, int targetSum, List<List<Integer>> ans,List<Integer> curr) {
        if(root == null) return;
        
        if(root.left == null && root.right == null) {
            if( targetSum-root.val==0){
              ArrayList<Integer> base = new ArrayList<>(curr);
                base.add(root.val);
                ans.add(base);   
            }
            return;
        }
        curr.add(root.val);
        
        pathSum(root.left,targetSum-root.val,ans,curr);
        pathSum(root.right,targetSum-root.val,ans,curr);
        
        curr.remove(curr.size()-1);
    }
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<List<Integer>> ans = new ArrayList<>();
        pathSum(root,targetSum,ans,new ArrayList<>());
        return ans;
    }
}

File: 0113-path-sum-ii/0113-path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
       void pathSum(TreeNode* root, int targetSum,vector<vector<int>> & ans,vector<int>&curr) {
        if(root == NULL) return;
        
        if(root->left == NULL && root->right == NULL) {
            if( targetSum-root->val==0){
              vector<int> base = curr;
                base.push_back(root->val);
                ans.push_back(base);   
            }
            return;
        }
        curr.push_back(root->val); //backtracking in curr
        
        pathSum(root->left,targetSum-root->val,ans,curr);
        pathSum(root->right,targetSum-root->val,ans,curr);
        
        curr.pop_back(); //backtracking in curr
    }
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<vector<int>> ans;
        vector<int> curr;
        pathSum(root,targetSum,ans,curr);
        return ans;
    }
};